### DISCUSSION TITLE
`minimumReleaseAge` being ignored for branch and PR creation

### DISCUSSION THREAD

**Discussion by MIJOTHY-V2** [renovatebot/renovate#32177] [Request Help] [answered]
### How are you running Renovate?

A Mend.io-hosted app

### Please tell us more about your question or problem

We've noticed our setting of `"minimumReleaseAge": "30 days"` has started being ignored for branch and PR creation. For example, based on our "monday mornings" schedule, today renovate created a branch and opened a PR for the following dependencies:

> | Package | Type | Update | Change |
> |---|---|---|---|
> | [ch.qos.logback:logback-classic](http://logback.qos.ch) ([source](https://redirect.github.com/qos-ch/logback)) | dependencies | patch | `1.5.11` -> `1.5.12` |
> | [ch.qos.logback:logback-core](http://logback.qos.ch) ([source](https://redirect.github.com/qos-ch/logback)) | dependencies | patch | `1.5.11` -> `1.5.12` |
> | [org.testcontainers:localstack](https://java.testcontainers.org) ([source](https://redirect.github.com/testcontainers/testcontainers-java)) | managed-dependencies | patch | `1.20.2` -> `1.20.3` |
> | [org.testcontainers:postgresql](https://java.testcontainers.org) ([source](https://redirect.github.com/testcontainers/testcontainers-java)) | managed-dependencies | patch | `1.20.2` -> `1.20.3` |
> | [org.testcontainers:testcontainers](https://java.testcontainers.org) ([source](https://redirect.github.com/testcontainers/testcontainers-java)) | managed-dependencies | patch | `1.20.2` -> `1.20.3` |

When we look at the [job logs](https://developer.mend.io/github/droitfintech/nuvo/-/job/0192d06c-57d1-7a5e-b918-68bbe63cf8a2) for this run (snippets provided below), we can see that Renovate acknowledges the minimum release age, so none of the above updates should have been proposed:

<details><summary>Release Age Logs</summary>

```
DEBUG: Update has not passed minimum release age (branch="renovate/clojure-dependencies")
{
  "depName": "ch.qos.logback:logback-classic"
  "timeElapsed": 186803228
  "minimumReleaseAge": "30 days"
}

DEBUG: Update has not passed minimum release age (branch="renovate/clojure-dependencies")
{
  "depName": "ch.qos.logback:logback-core"
  "timeElapsed": 186833229
  "minimumReleaseAge": "30 days"
}

DEBUG: Update has not passed minimum release age (branch="renovate/clojure-dependencies")
{
  "depName": "org.testcontainers:localstack"
  "timeElapsed": 444555229
  "minimumReleaseAge": "30 days"
}

DEBUG: Update has not passed minimum release age (branch="renovate/clojure-dependencies")
{
  "depName": "org.testcontainers:postgresql"
  "timeElapsed": 444508229
  "minimumReleaseAge": "30 days"
}

DEBUG: Update has not passed minimum release age (branch="renovate/clojure-dependencies")
{
  "depName": "org.testcontainers:testcontainers"
  "timeElapsed": 444435229
  "minimumReleaseAge": "30 days"
}
```

</details>

Furthermore, once the branch and PR are incorrectly created, the `renovate/stability-days` status check shows that the updates have not passed the relevant release age: 
![image](https://github.com/user-attachments/assets/27cd8f58-6643-4dac-89c9-e8fb45b7cffa)

This has been happening across multiple of our repositories, and seems to be a relatively recent issue, as we tested `minimumReleaseAge` behaviour in July and August. The Renovate app version in the logs we see today is `38.120.1`.

### Logs (if relevant)

<details><summary>Release Age Logs</summary>

```
DEBUG: Update has not passed minimum release age (branch="renovate/clojure-dependencies")
{
  "depName": "ch.qos.logback:logback-classic"
  "timeElapsed": 186803228
  "minimumReleaseAge": "30 days"
}

DEBUG: Update has not passed minimum release age (branch="renovate/clojure-dependencies")
{
  "depName": "ch.qos.logback:logback-core"
  "timeElapsed": 186833229
  "minimumReleaseAge": "30 days"
}

DEBUG: Update has not passed minimum release age (branch="renovate/clojure-dependencies")
{
  "depName": "org.testcontainers:localstack"
  "timeElapsed": 444555229
  "minimumReleaseAge": "30 days"
}

DEBUG: Update has not passed minimum release age (branch="renovate/clojure-dependencies")
{
  "depName": "org.testcontainers:postgresql"
  "timeElapsed": 444508229
  "minimumReleaseAge": "30 days"
}

DEBUG: Update has not passed minimum release age (branch="renovate/clojure-dependencies")
{
  "depName": "org.testcontainers:testcontainers"
  "timeElapsed": 444435229
  "minimumReleaseAge": "30 days"
}
```

</details>

<details><summary>Package File Update Logs</summary>

```
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=false (branch="renovate/clojure-dependencies")
DEBUG: Starting search at index 2498 (branch="renovate/clojure-dependencies")
{
  "packageFile": "adept-java/project.clj"
  "depName": "ch.qos.logback:logback-classic"
}

DEBUG: Found match at index 2498 (branch="renovate/clojure-dependencies")
{
  "packageFile": "adept-java/project.clj"
  "depName": "ch.qos.logback:logback-classic"
}

DEBUG: Contents updated (branch="renovate/clojure-dependencies")
{
  "packageFile": "adept-java/project.clj"
  "depName": "ch.qos.logback:logback-classic"
}

DEBUG: Value is not updated (branch="renovate/clojure-dependencies")
{
  "depName": "ch.qos.logback:logback-core"
  "manager": "leiningen"
  "packageFile": "adept-java/project.clj"
  "expectedValue": "1.5.12"
  "foundValue": "1.5.11"
}
```

</details>

<details><summary>Branch/PR Creation Logs</summary>

```
EBUG: latest commit (branch="renovate/clojure-dependencies")
{
  "branchName": "master"
  "latestCommitDate": "2024-10-25T14:26:27+01:00"
}

INFO: Branch created (branch="renovate/clojure-dependencies")
{
  "commitSha": "04f71607c83ad3a745daa81decd4d4f89e516f79"
}

DEBUG: Updating renovate/stability-days status check state to yellow (branch="renovate/clojure-dependencies")
DEBUG: Setting branch status (branch="renovate/clojure-dependencies")
{
  "context": "renovate/stability-days"
  "state": "yellow"
}

DEBUG: http cache: saving https://api.github.com/repos/droitfintech/nuvo/commits/renovate/clojure-dependencies/status (etag=W/"acd63abdbac5ef1ae6f8525c8096abe6dfe3e74dc7063cafea1ccf957e47e6f7", lastModified=undefined) (branch="renovate/clojure-dependencies")
DEBUG: Ensuring PR (branch="renovate/clojure-dependencies")
DEBUG: There are 0 errors and 0 warnings (branch="renovate/clojure-dependencies")
DEBUG: getBranchPr(renovate/clojure-dependencies) (branch="renovate/clojure-dependencies")
DEBUG: findPr(renovate/clojure-dependencies, undefined, open) (branch="renovate/clojure-dependencies")
DEBUG: findPr(renovate/clojure-dependencies, undefined, closed) (branch="renovate/clojure-dependencies")
DEBUG: Fetching changelog: https://github.com/qos-ch/logback (1.5.11 -> 1.5.12) (branch="renovate/clojure-dependencies")
DEBUG: Fetching changelog: https://github.com/qos-ch/logback (1.5.11 -> 1.5.12) (branch="renovate/clojure-dependencies")
DEBUG: Fetching changelog: https://github.com/testcontainers/testcontainers-java (1.20.2 -> 1.20.3) (branch="renovate/clojure-dependencies")
DEBUG: Fetching changelog: https://github.com/testcontainers/testcontainers-java (1.20.2 -> 1.20.3) (branch="renovate/clojure-dependencies")
DEBUG: Fetching changelog: https://github.com/testcontainers/testcontainers-java (1.20.2 -> 1.20.3) (branch="renovate/clojure-dependencies")
DEBUG: Creating PR (branch="renovate/clojure-dependencies")
{
  "prTitle": "Update Clojure dependencies"
}

DEBUG: Creating PR (branch="renovate/clojure-dependencies")
{
  "title": "Update Clojure dependencies"
  "head": "droitfintech:renovate/clojure-dependencies"
  "base": "master"
  "draft": false
}

DEBUG: PR created (branch="renovate/clojure-dependencies")
{
  "pr": 6489
  "draft": false
}

DEBUG: Adding labels '' to #6489 (branch="renovate/clojure-dependencies")
DEBUG: GitHub-native automerge: success...PrNo: 6489 (branch="renovate/clojure-dependencies")
INFO: PR created (branch="renovate/clojure-dependencies")
{
  "pr": 6489
  "prTitle": "Update Clojure dependencies"
}

DEBUG: resolveBranchStatus(branchName=renovate/clojure-dependencies, ignoreTests=false) (branch="renovate/clojure-dependencies")
DEBUG: getBranchStatus(renovate/clojure-dependencies) (branch="renovate/clojure-dependencies")
DEBUG: http cache: Using cached response: https://api.github.com/repos/droitfintech/nuvo/commits/renovate/clojure-dependencies/status from 2024-10-28T00:36:17.372Z (branch="renovate/clojure-dependencies")
DEBUG: branch status check result (branch="renovate/clojure-dependencies")
{
  "state": "pending"
  "statuses": [
    {
      "url": "https://api.github.com/repos/droitfintech/nuvo/statuses/04f71607c83ad3a745daa81decd4d4f89e516f79",
      "avatar_url": "https://avatars.githubusercontent.com/in/2740?v=4",
      "id": 32222859222,
      "node_id": "SC_kwDOAyXAtM8AAAAHgKHP1g",
      "state": "pending",
      "description": "Updates have not met minimum release age requirement",
      "target_url": "https://docs.renovatebot.com/configuration-options#minimumreleaseage",
      "context": "renovate/stability-days",
      "created_at": "2024-10-28T00:36:16Z",
      "updated_at": "2024-10-28T00:36:16Z"
    }
  ]
}

DEBUG: No check runs found (branch="renovate/clojure-dependencies")
{
  "result": {
    "total_count": 0,
    "check_runs": []
  }
}

DEBUG: Branch status yellow (branch="renovate/clojure-dependencies")
```

</details>

**Comment by MIJOTHY-V2** [comment]
I have produced a minimum repro here: https://github.com/MIJOTHY-V2/renovate-32177. The repro also includes the entire repository job logs for the renovate run that created the PR (vs the snippets included in this issue).

**Reply by RahulGautamSingh** [reply]
Try this: https://docs.renovatebot.com/configuration-options/#prcreation

 I expect what you are facing to be the intended behaviour. But if you think behaviour has changed recently, without you changing your config then I will check in a while, if there's been any changes.

**Reply by MIJOTHY-V2** [reply]
> I expect what you are facing to be the intended behaviour. But if you think behaviour has changed recently, without you changing your config then I will check in a while, if there's been any changes.

Yeh I am fairly sure this is a change in behaviour rather than something that's always occurred. Per the docs [here](https://docs.renovatebot.com/configuration-options/#suppress-branchpr-creation-for-x-days):

> If you use minimumReleaseAge=3 days and internalChecksFilter="strict" then Renovate only creates branches when 3 (or more days) have passed since the version was released. We recommend you set dependencyDashboard=true, so you can see these pending PRs.

the branch should never be created in the first place, using the configuration we have. The `prCreation` option should not be a factor in the behaviour we are seeing, where both the branch _and_ the PR are being created unexpectedly. Changing `prCreation` would still result in the branch being created ahead of the `minimumReleaseAge` check being satisfied, which is undesired behaviour in our case.

I've described the issue incompletely as I've only referenced "PR creation", which is misleading. I've now updated the issue and repro README to speak about both branch and PR creation. So to clarify the expected behaviour:

With a configuration like

```json
{
    "minimumReleaseAge": "30 days",
    "internalChecksFilter": "strict",
    "dependencyDashboard": true
}
```

and using the default `prCreation` value of `immediate`, I would expect Renovate to 
* Only create a branch for an update, once the update passes the `minimumReleaseAge` criteria
* Create a PR for an update, as soon as the corresponding branch has been created

**Comment by WeisSebastian** [comment]
We also noticed this on our repo. We didn't change anything in the config, but suddenly renovate started creating PR's even when our minimumReleaseAge of 7 days didn't expires. Interestingly it seems to respect it sometimes?
We do not have prCreation set like @RahulGautamSingh suggests, but we do have [internalChecksFilter](https://docs.renovatebot.com/configuration-options/#internalchecksfilter) set to strict

**Reply by leonardehrenfried** [reply]
I've started seeing the same and we have neither option in our config file.

**Comment by rarkins** [comment, marked as answer]
This is hopefully fixed by #32204

**Reply by MIJOTHY-V2** [reply]
Tried this out in the repro repo and it was indeed fixed. I've updated the [repro README](https://github.com/MIJOTHY-V2/renovate-32177?tab=readme-ov-file#32177) accordingly.

Thanks for the quick turnaround, it's really appreciated.

**Reply by leonardehrenfried** [reply]
I agree, this is impressive work.
